modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java

```diff
@@ -4190,4 +4224,150 @@ public boolean addTimeoutHandler() {
             "mappings", mappings,
             "super", super.toString());
     }
+
+    /**
+     * Creates savepoint.
+     *
+     * @param name Savepoint ID.
+     * @param overwrite Overwrite savepoint flag.
+     */
+    @Override public void savepoint(String name, boolean overwrite) throws IgniteCheckedException {
+        txState.awaitLastFuture(cctx);
+
+        checkValid();
+
+        if (state() != ACTIVE) {
+            throw new IllegalStateException("Trying to create savepoint not in ACTIVE transaction." +
+                " Savepoint will not be created.");
+        }
+
+        ListIterator<TxSavepointLocal> spIter = findSP(name);
+
+        if (spIter != null) {
+            if (overwrite) {
+                releaseSP(spIter);
+            } else {
+                throw new IllegalArgumentException("Savepoint \"" + name + "\" already exist. " +
+                    "Savepoints with the same name aren't available. " +
+                    "If you want to rewrite savepoints - use savepoint(name, true) method.");
+            }
+        }
+
+        addSP(new TxSavepointLocal(name, this));
+
+        if (log.isDebugEnabled())
+            log.debug("Saving point created [savepoint=" + name + ", tx=" + this + ']');
+    }
+
+    /**
+     * Returns transaction to previously saved state.
+     *
+     * @param name Savepoint ID.
+     */
+    @Override public void rollbackToSavepoint(String name) throws IgniteCheckedException {
+        txState.awaitLastFuture(cctx);
+
+        checkValid();
+
+        if (state() != ACTIVE) {
+            throw new IllegalStateException("Trying to rollback to savepoint not in ACTIVE transaction." +
+                " Nothing will be rolled back.");
+        }
+
+        ListIterator<TxSavepointLocal> spIter = findSP(name);
+
+        if (spIter == null)
+            throw new IllegalArgumentException("No such savepoint.");
+
+        txState.rollbackToSavepoint(spIter.next(), cctx, this);
+
+        releaseSPs(spIter);
+
+        if (log.isDebugEnabled())
+            log.debug("Rolled back to savepoint [savepoint=" + name + ", tx=" + this + ']');
+    }
+
+    /**
+     * Release savepoint and remove it from savepoint list. Rallback to this savepoint will be unavailable.
+     *
+     * @param name Savepoint ID.
+     */
+    @Override public void releaseSavepoint(String name) throws IgniteCheckedException {
+        txState.awaitLastFuture(cctx);
+
+        checkValid();
+
+        if (state() != ACTIVE) {
+            throw new IllegalStateException("Trying to release savepoint not in ACTIVE transaction." +
```

Comment:
"Trying to release savepoint not in ACTIVE transaction." -> "Trying to release savepoint not in " + ACTIVE + " transaction."